### PR TITLE
Add complex case mapping and title case mapping.

### PR DESCRIPTION
**Pull request by SimonSapin**
- Add “complex” mappings to `char::to_lowercase` and `char::to_uppercase`, making them yield sometimes more than on `char`: #25800. `str::to_lowercase` and `str::to_uppercase` are affected as well.
- Add `char::to_titlecase`, since it’s the same algorithm (just different data). However this does **not** add `str::to_titlecase`, as that would require UAX#29 Unicode Text Segmentation which we decided not to include in of `std`: https://github.com/rust-lang/rfcs/pull/1054 I made `char::to_titlecase` immediately `#[stable]`, since it’s so similar to `char::to_uppercase` that’s already stable. Let me know if it should be `#[unstable]` for a while.
- Add a special case for upper-case Sigma in word-final position in `str::to_lowercase`: #26035. This is the only language-independent conditional mapping currently in `SpecialCasing.txt`.
- Stabilize `str::to_lowercase` and `str::to_uppercase`. The `&self -> String` on `str` signature seems straightforward enough, and the only relevant issue I’ve found is #24536 about naming. But `char` already has stable methods with the same name, and deprecating them for a rename doesn’t seem worth it.

r? @alexcrichton

### CODE REVIEW

**Comment by SimonSapin** [comment]
Regarding the table sizes: there used to be two `&'static [(char, char)]` tables of 1100 and 1092 items, for uppercase and lowercase mapping respectively, for a total of 17 536 bytes. These are now `&'static [(char, [char; 3])]` with 1175 and 1092 entries, so 36 272 bytes. Adding another 1152 entries table for titlecase brings the total to 54 704 bytes.

It’s probably possible to trade off some CPU time in mapping execution to make the table smaller: most of the `[char; 3]` arrays are `[_, 0, 0]` (indicating a mapping to a single `char`), and the titlecase table has a large common subset with the uppercase one.

File: src/libcollections/str.rs [hunk 1]
```diff
@@ -1851,11 +1851,35 @@ impl str {
...
     /// assert_eq!(s.to_lowercase(), "hello");
     /// ```
-    #[unstable(feature = "collections")]
+    #[stable(feature = "unicode_case_mapping", since = "1.2.0")]
     pub fn to_lowercase(&self) -> String {
         let mut s = String::with_capacity(self.len());
-        s.extend(self[..].chars().flat_map(|c| c.to_lowercase()));
+        for (i, c) in self[..].char_indices() {
+            if c == 'Σ' {
+                map_uppercase_sigma(self, i, &mut s)
```
Comment (alexcrichton) [inline src/libcollections/str.rs]: Could you add a comment for why this is being specially cased here?

File: src/libcollections/str.rs [hunk 2]
```diff
@@ -1851,11 +1851,35 @@ impl str {
...
+            } else {
+                s.extend(c.to_lowercase());
+            }
+        }
         return s;
+
+        #[cold]
+        #[inline(never)]
```
Comment (alexcrichton) [inline src/libcollections/str.rs]: Were you able to get a measurable improvement with these attributes? If not I'd personally recommend leaving them off for now.

File: src/libcollectionstest/str.rs [hunk 3]
```diff
@@ -1687,6 +1687,19 @@ fn trim_ws() {
                "");
 }
 
+#[test]
+fn to_lowercase() {
+    assert_eq!("".to_lowercase(), "");
+    // https://github.com/rust-lang/rust/issues/26035
+    assert_eq!("'Σ AÉΣ'Σ'' Σ ǅΣ".to_lowercase(), "'σ aéσ'ς'' σ ǆς");
```
Comment (alexcrichton) [inline src/libcollectionstest/str.rs]: There's a few other conditions in the checking for the sigma at the end, could this add some tests cases for those?

File: src/librustc_unicode/char.rs [hunk 4]
```diff
@@ -397,27 +452,48 @@ impl char {
...
+    /// digraphs and ligature characters.
+    /// For example, U+01F3 “ǳ” and U+FB01 “ﬁ”
+    /// map to U+01F1 “Ǳ” and U+0046 U+0069 “Fi”, respectively.
+    ///
+    /// # Return value
+    ///
+    /// Returns an iterator which yields the characters corresponding to the
+    /// lowercase equivalent of the character. If no conversion is possible then
+    /// an iterator with just the input character is returned.
+    #[stable(feature = "unicode_case_mapping", since = "1.2.0")]
```
Comment (alexcrichton) [inline src/librustc_unicode/char.rs]: I think this should start off as `#[unstable]` as it's a new API

**Comment by alexcrichton** [comment]
I'm ok adding the extra table sizes for now, with `--gc-sections` it should get rid of unused ones anyway and we can always explore a more optimized representation later.

**Comment by alexcrichton** [comment]
cc @rust-lang/libs

File: src/libcollections/str.rs (hunk 2 above)
Comment (SimonSapin) [inline src/libcollections/str.rs]: I did not take any measurement, I just imitated `Vec::push`. I’ll remove them.

File: src/libcollectionstest/str.rs (hunk 3 above)
Comment (SimonSapin) [inline src/libcollectionstest/str.rs]: What cases are you thinking of? End of a word, of the string?

File: src/libcollectionstest/lib.rs [hunk 5]
```diff
@@ -37,6 +37,7 @@ extern crate rustc_unicode;
 mod binary_heap;
 mod bit;
 mod btree;
+mod char;  // char isn't really a collection, but didn't find a better place for this.
```
Comment (Gankro) [inline src/libcollectionstest/lib.rs]: mumble mumble collection of code points mumble mumble

File: src/libcollectionstest/str.rs (hunk 3 above)
Comment (alexcrichton) [inline src/libcollectionstest/str.rs]: Ah specifically there's a boolean `is_word_final` and this only exercises one of the two code paths (e.g. only the `true` case I believe)

File: src/libcollectionstest/lib.rs (hunk 5 above)
Comment (SimonSapin) [inline src/libcollectionstest/lib.rs]: A `char` is *one* code point.

File: src/libcollectionstest/str.rs (hunk 3 above)
Comment (SimonSapin) [inline src/libcollectionstest/str.rs]: All `σ`’s come from the false case, but I wrote more tests anyway.

**Comment by SimonSapin** [comment]
Updated.

**Comment by alexcrichton** [comment]
@bors: r+ c57a4124ff6eeb848a537d0591235fa6e56f1cf7

Thanks @SimonSapin!

**Comment by SimonSapin** [comment]
```
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

That’s not because of my patch, is it?

**Comment by SimonSapin** [comment]
Turns out `char::to_{upp,low}ercase` *did* have tests already. I hadn’t found them as I didn’t suspect `libcoretest` would depend on `librustc_unicode`.

This should fix it.

**Comment by SimonSapin** [comment]
I’m getting an LLVM assertion error in `make check` locally :(

```
---- [run-pass] run-pass/sepcomp-lib-lto.rs stdout ----
	
error: compilation failed!
status: signal: 6
command: x86_64-unknown-linux-gnu/stage2/bin/rustc /home/simon/projects/rust/src/test/run-pass/sepcomp-lib-lto.rs -L x86_64-unknown-linux-gnu/test/run-pass/ --target=x86_64-unknown-linux-gnu -L x86_64-unknown-linux-gnu/test/run-pass/sepcomp-lib-lto.stage2-x86_64-unknown-linux-gnu.run-pass.libaux -o x86_64-unknown-linux-gnu/test/run-pass/sepcomp-lib-lto.stage2-x86_64-unknown-linux-gnu --cfg rtopt -O -L x86_64-unknown-linux-gnu/rt -C lto
stdout:
------------------------------------------
 
------------------------------------------
stderr:
------------------------------------------
rustc: ../../../../../../projects/rust/src/llvm/lib/CodeGen/LexicalScopes.cpp:179: llvm::LexicalScope* llvm::LexicalScopes::getOrCreateRegularScope(llvm::MDNode*): Assertion `DISubprogram(Scope).describes(MF->getFunction())' failed.
 
------------------------------------------

**Comment by SimonSapin** [comment]
```
command timed out: 1200 seconds without output, attempting to kill
```

**Comment by brson** [comment]
This will need to be mentioned in the 1.2 release notes and it's a breaking change.

**Comment by SimonSapin** [comment]
Is it? I don’t really know how to interpret https://github.com/rust-lang/rfcs/blob/master/text/1105-api-evolution.md#behavioral-changes here, but the new behaviour in this PR is the one that has been documented as intended since before 1.0.

**Comment by brson** [comment]
@SimonSapin It's allowed, but it changes behavior and so can break users' code (regardless of what the policy may consider 'breaking').

**Comment by SimonSapin** [comment]
Ok. Agreed on mentioning in the release notes.

**Comment by brson** [comment]
I tagged this with stable-regression since it's a change in behavior and that's currently the best tag to track them, even though there is definitely a difference between intentional and unintentional regressions.

**Pull request merged**